client/extensions/woocommerce/lib/generate-variations/index.js

```diff
@@ -0,0 +1,36 @@
+// http://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
+const f = ( a, b ) => [].concat( ...a.map( c => b.map( d => [].concat( c, d ) ) ) );
+const cartesian = ( a, b, ...c ) => b ? cartesian( f( a, b ), ...c ) : a;
```

Comment:
Also, it probably will be more performant to construct first the `{ name:"Size", option:"Small" }` objects, then do the `cartesian product` over those whole objects. The objects won't be copied, so the RAM footprint of the (possibly massive) final array will be smaller. Plus, you won't need to iterate over the array again to add the variation names.